api/apps/articles/model/save.coffee

```diff
@@ -85,6 +85,40 @@ sailthru = require('sailthru-client').createSailthruClient(SAILTHRU_KEY,SAILTHRU
     article.keywords = keywords[0..9]
     cb(null, article)
 
+@generateArtworks = (article, accessToken, input, cb) ->
+  # First check if any sections have artworks
+  return cb(null, article) unless _.some input.sections, type: 'artworks'
+  callbacks = []
+  # Try to fetch and denormalize the artworks from Gravity asynchonously
+  for section in input.sections when section.type is 'artworks'
+    for artworkId in section.ids
+      do (artworkId) ->
+        callbacks.push (callback) ->
+          request
+            .get("#{ARTSY_URL}/api/v1/artwork/#{artworkId}")
+            .set('X-Xapp-Token': accessToken)
+            .end callback
+  async.parallel callbacks, (err, results) =>
+    return cb(err) if err
+    fetchedArtworks = results.map (result) -> result.body
+    # Push fetched artworks to the section if they are available
+    for section in input.sections when section.type is 'artworks'
+      section.artworks = []
+      for artworkId in section.ids
+        artwork = _.findWhere fetchedArtworks, _id: artworkId
+        if artwork
+          section.artworks.push denormalizedArtworkData artwork
+        else
+          section.ids = _.without section.ids, artworkId
+      input.sections = _.without input.sections, section if section.ids.length is 0
```

Comment:
I think it would be more appropriate, readability-wise, and safer* to not mutate `input.sections`. Instead I think it would be better to build a new sections array. Wouldn't change much code—mostly `for section in newSections = _.clone(input.sections)` and `article.sections = newSections`.

* This can help prevent issues down the line where someone expects `input` to consistently be the data being saved to the article, and not having that change underneath them. Functional languages like Clojure/Elxir use [immutable](https://facebook.github.io/immutable-js/) data structures precisely b/c of pitfalls like this where it's easy to have some object change state from underneath you.